File: android/app/src/main/java/com/cashu/me/Models/Transactions/WalletTransaction.kt
package com.cashu.me.Models

import kotlinx.serialization.Serializable
import com.cashu.me.Core.PaymentRequestDecoder

@Serializable
data class WalletTransaction(
    val id: String,
    val amount: Long,
    val type: TransactionType,
    val kind: TransactionKind,
    val dateEpochMillis: Long,
    val memo: String? = null,
    val status: TransactionStatus,
    val statusNote: String? = null,
    val mintUrl: String? = null,
    val preimage: String? = null,
    val token: String? = null,
    val invoice: String? = null,
    val fee: Long = 0,
    /** Mint account unit for [amount] and [fee] (sat, usd, eur, or custom). */
    val unit: String = "sat",
    /**
     * CDK wallet-saga (operation) id backing this transaction, when the row
     * came from CDK. Pending sent tokens use it for claim checks / revoke;
     * null for app-synthesized rows (quotes, held receives).
     */
    val sagaId: String? = null,
    val quoteId: String? = null,
    val cashuRequestId: String? = null,
    /** Incoming ecash held for user approval ("Receive later" / NUT-18). Opens
     * the claim flow instead of a plain receipt. iOS parity. */
    val isPendingReceiveToken: Boolean = false,
    /** BOLT11 mint quote still awaiting payment — titles the row "Lightning invoice". */
    val isUnpaidInvoice: Boolean = false,
) {
    val displayDescription: String?
        get() = memo?.takeIf(String::isNotBlank) ?: PaymentRequestDecoder.description(invoice)

    /** Display the decoder's hashed Lightning description as a copyable reference. */
    val descriptionHash: String?
        get() {
            if (kind != TransactionKind.Lightning) return null
            val description = displayDescription?.trim() ?: return null
            if (!description.startsWith("Hash:")) return null
            return description.removePrefix("Hash:").trim().takeIf { hash ->
                hash.length == 64 && hash.all { it in '0'..'9' || it in 'a'..'f' || it in 'A'..'F' }
            }
        }

    val displayStatusText: String
        get() = if (status == TransactionStatus.Pending) statusNote ?: status.displayText else status.displayText

    /** Quiet Pending rule: expired never credited the balance, so it keeps the bare muted amount. */
    val isUnsettled: Boolean
        get() = status == TransactionStatus.Pending || status == TransactionStatus.Expired

    /**
     * Mint-quote id to re-check when opening this row's detail, if any.
     * Incoming unsettled mint quotes and reusable offers (not ecash or melts).
     * Expired unpaid invoices are included so a late-paid NUT-04
     * quote can still mint after the invoice timer.
     */
    val mintQuoteIdForStatusRefresh: String?
        get() {
            if (type != TransactionType.Incoming) return null
            if (kind != TransactionKind.Lightning && kind != TransactionKind.Onchain) return null
            if (invoice == null) return null
            val reusableOffer = kind == TransactionKind.Lightning &&
                status == TransactionStatus.Completed && invoice.startsWith("lno", ignoreCase = true)
            if (status != TransactionStatus.Pending && status != TransactionStatus.Expired && !reusableOffer) return null
            return quoteId ?: id
        }
}

@Serializable
enum class TransactionType {
    Incoming,
    Outgoing,
}

@Serializable
enum class TransactionKind {
    Ecash,
    Lightning,
    Onchain;

    val displayName: String
        get() = when (this) {
            Ecash -> "Ecash"
            Lightning -> "Lightning"
            Onchain -> "On-chain"
        }
}

@Serializable
enum class TransactionStatus {
    Pending,
    Completed,
    Failed,
    Expired;

    val displayText: String
        get() = when (this) {
            Pending -> "Pending"
            Completed -> "Completed"
            Failed -> "Failed"
            Expired -> "Expired"
        }
}

/**
 * Resolve the live row for a detail screen opened with [openId] (and the
 * open-time [openQuoteId]). Pending quote rows use `id == quoteId`; once
 * minting starts CDK replaces them with a saga-derived transaction id that
 * still carries `quoteId`, so fall back to the quoteId to keep following the
 * row as it flips Pending → Completed in place. Rows are newest-first, so a
 * reusable offer resolves to its latest payment. iOS `liveDetail` parity.
 */
internal fun List<WalletTransaction>.liveDetail(
    openId: String,
    openQuoteId: String? = null,
): WalletTransaction? =
    firstOrNull { it.id == openId }
        ?: firstOrNull { it.quoteId != null && it.quoteId == openQuoteId }

/** CDK may omit a mint transaction's memo and request after settlement. */
internal fun WalletTransaction.restoringDescription(requests: List<CashuRequest>): WalletTransaction {
    val description = displayDescription ?: requests.firstOrNull { request ->
        type == TransactionType.Incoming && unit.equals(request.unit, ignoreCase = true) &&
            (request.receivedPayments.any { it.transactionId == id } || cashuRequestId == request.id ||
                (quoteId != null && quoteId == request.quoteId &&
                    request.mints.any { it.trimEnd('/') == mintUrl?.trimEnd('/') }))
    }?.displayDescription
    return if (description == memo) this else copy(memo = description)
}

File: ios/CashuWallet/Models/Transactions/WalletTransaction.swift
import Foundation

struct WalletTransaction: Identifiable {
    let id: String
    let amount: UInt64
    let type: TransactionType
    let kind: TransactionKind
    let date: Date
    var memo: String?

    var displayDescription: String? {
        if let memo, !memo.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return memo }
        return PaymentRequestDecoder.description(from: invoice)
    }

    /// The decoder can expose a hashed Lightning description as `Hash: <hex>`.
    /// Keep the stored description intact, but display this reference as a copyable row.
    var descriptionHash: String? {
        guard kind == .lightning,
              let description = displayDescription?.trimmingCharacters(in: .whitespacesAndNewlines),
              description.hasPrefix("Hash:") else { return nil }
        let hash = description.dropFirst(5).trimmingCharacters(in: .whitespacesAndNewlines)
        guard hash.utf8.count == 64,
              hash.utf8.allSatisfy({ (48...57).contains($0) || (65...70).contains($0) || (97...102).contains($0) })
        else { return nil }
        return hash
    }

    var status: TransactionStatus
    var statusNote: String? = nil
    
    /// Associated mint URL
    var mintUrl: String?
    
    /// Payment proof (preimage for Lightning, txid for on-chain when exposed)
    var preimage: String?
    
    /// Ecash token string (for outgoing pending transactions)
    var token: String?
    
    /// Payment request string (BOLT11 invoice, BOLT12 offer, or on-chain address)
    var invoice: String?
    
    /// Fee paid for the transaction (in sats)
    var fee: UInt64 = 0

    /// Mint account unit for `amount` ("sat", "usd", "eur", or custom).
    var unit: String = "sat"

    /// CDK wallet-saga (operation) id backing this transaction, when the row
    /// came from CDK. Pending sent tokens use it for `checkSendStatus` /
    /// `revokeSend`; nil for app-synthesized rows (quotes, held receives).
    var sagaId: String? = nil

    /// Incoming ecash the user hasn't claimed yet (a "Receive Later" token or
    /// a NUT-18 payment held for approval). Its receipt offers the claim flow.
    var isPendingReceiveToken: Bool = false

    /// Source Cashu Request id when this incoming ecash transaction was
    /// auto-claimed via NUT-18. History uses this to suppress the duplicate
    /// row in favor of the request row.
    var cashuRequestId: String? = nil

    /// Mint-quote id for Lightning / on-chain mints and melts. The join key
    /// used to attach this transaction to the receive-intent backing its quote
    /// (a reusable BOLT12 offer, a BOLT11 invoice, an on-chain address).
    var quoteId: String? = nil

    /// BOLT11 mint quote still awaiting payment — titles the row
    /// "Lightning invoice" until the invoice settles.
    var isUnpaidInvoice: Bool = false

    /// The Quiet Pending treatment (bare, muted amount) covers expired too:
    /// an expired invoice never credited the balance.
    var isUnsettled: Bool {
        status == .pending || status == .expired
    }

    /// Mint-quote id to re-check when opening this row's detail, if any.
    /// Incoming unsettled mint quotes and reusable offers (not ecash or melts).
    /// Expired unpaid invoices are included so a late-paid NUT-04
    /// quote can still mint after the invoice timer.
    var mintQuoteIdForStatusRefresh: String? {
        guard type == .incoming else { return nil }
        guard kind == .lightning || kind == .onchain else { return nil }
        guard !isPendingReceiveToken else { return nil }
        guard invoice != nil else { return nil }
        let reusableOffer = kind == .lightning && status == .completed &&
            invoice?.lowercased().hasPrefix("lno") == true
        guard status == .pending || status == .expired || reusableOffer else { return nil }
        return quoteId ?? id
    }

    /// Payment receipts retire their QR after settlement. Reusable offers remain
    /// available from their dedicated request detail screen.
    var hasActionablePaymentCode: Bool {
        switch kind {
        case .ecash:
            return type == .outgoing && status == .pending && token?.isEmpty == false
        case .lightning:
            guard let invoice, !invoice.isEmpty else { return false }
            return status == .pending
        case .onchain:
            return status == .pending && invoice?.isEmpty == false
        }
    }

    var displayStatusText: String {
        if status == .pending {
            return statusNote ?? status.displayText
        }

        return status.displayText
    }

    /// Canonical row/detail title — kind-first, capitalized kind, lowercase
    /// verb, parallel across all kinds (e.g. "Ecash received", "Lightning
    /// paid"). Single source of truth for the History/Home rows and the
    /// transaction detail nav title.
    var displayTitle: String {
        if isPendingReceiveToken { return "Ecash to claim" }
        // Nothing has been received while the invoice awaits payment.
        if isUnpaidInvoice { return "Lightning invoice" }
        switch (kind, type) {
        case (.ecash,     .incoming): return "Ecash received"
        case (.ecash,     .outgoing): return "Ecash sent"
        case (.lightning, .incoming): return "Lightning received"
        case (.lightning, .outgoing): return "Lightning paid"
        case (.onchain,   .incoming): return "Bitcoin received"
        case (.onchain,   .outgoing): return "Bitcoin sent"
        }
    }

    enum TransactionType {
        case incoming   // Mint or receive
        case outgoing   // Send or melt
        
        var icon: String {
            switch self {
            case .incoming: return "arrow.down.circle.fill"
            case .outgoing: return "arrow.up.circle.fill"
            }
        }
    }
    
    /// Kind of transaction - distinguishes between Ecash and Lightning
    enum TransactionKind {
        case ecash      // Ecash token send/receive
        case lightning  // Lightning invoice mint/melt
        case onchain    // On-chain address mint/melt
        
        var displayName: String {
            switch self {
            case .ecash: return "Ecash"
            case .lightning: return "Lightning"
            case .onchain: return "On-chain"
            }
        }
    }
    
    enum TransactionStatus {
        case pending
        case completed
        case failed
        case expired

        var displayText: String {
            switch self {
            case .pending: return "Pending"
            case .completed: return "Completed"
            case .failed: return "Failed"
            case .expired: return "Expired"
            }
        }
    }
}

extension Array where Element == WalletTransaction {
    /// Resolve the live row for a detail screen opened with `openId` (and the
    /// open-time `openQuoteId`). Pending quote rows use `id == quoteId`; once
    /// minting starts CDK replaces them with a saga-derived transaction id that
    /// still carries `quoteId`, so fall back to the quoteId to keep following
    /// the row as it flips Pending → Completed in place. Rows are newest-first,
    /// so a reusable offer resolves to its latest payment.
    func liveDetail(openId: String, openQuoteId: String? = nil) -> WalletTransaction? {
        first(where: { $0.id == openId })
            ?? first(where: { $0.quoteId != nil && $0.quoteId == openQuoteId })
    }
}

/// Home is a compact settled-ledger view, not an operational queue. Generated
/// receive artifacts and every non-completed state remain available in History.
enum HomeActivity {
    static func recentTransactions(
        from transactions: [WalletTransaction],
        limit: Int
    ) -> [WalletTransaction] {
        transactions
            .filter { $0.status == .completed }
            .sorted { $0.date > $1.date }
            .prefix(max(0, limit))
            .map { $0 }
    }
}

/// CDK derives the transaction id of a saga-managed operation from the
/// operation (UUID) id: the id's dash-free ASCII form becomes the 32 id bytes,
/// which the FFI then hex-encodes. Reproduced locally because the FFI helper
/// (`TransactionId.from_saga_id`) is not exported to the bindings.
enum SagaTransactionId {
    /// Operation (saga) UUID string → CDK transaction id hex.
    static func transactionIdHex(operationId: String) -> String? {
        let simple = operationId.lowercased().replacingOccurrences(of: "-", with: "")
        guard simple.count == 32, simple.allSatisfy({ $0.isHexDigit }) else { return nil }
        return Data(simple.utf8).map { String(format: "%02x", $0) }.joined()
    }

    /// CDK transaction id hex → operation (saga) UUID string, as accepted by
    /// `checkSendStatus` / `revokeSend` (UUID parsing tolerates the simple form).
    static func operationId(fromTransactionIdHex txId: String) -> String? {
        var bytes = [UInt8]()
        bytes.reserveCapacity(txId.count / 2)
        var index = txId.startIndex
        while index < txId.endIndex {
            let next = txId.index(index, offsetBy: 2, limitedBy: txId.endIndex) ?? txId.endIndex
            guard let byte = UInt8(txId[index..<next], radix: 16) else { return nil }
            bytes.append(byte)
            index = next
        }
        guard bytes.count == 32 else { return nil }
        let simple = String(decoding: bytes, as: UTF8.self)
        guard simple.count == 32, simple.allSatisfy({ $0.isHexDigit }) else { return nil }
        return simple
    }
}


extension WalletTransaction {
    /// CDK may omit a mint transaction's memo and request after settlement.
    func restoringDescription(from requests: [CashuRequest]) -> WalletTransaction {
        let request = requests.first { request in
            type == .incoming && unit.lowercased() == request.unit.lowercased() &&
                (request.receivedPayments.contains { $0.transactionId == id } || cashuRequestId == request.id ||
                    (quoteId != nil && quoteId == request.quoteId &&
                        request.mints.contains { MintURLIdentity.normalized($0) == mintUrl.map(MintURLIdentity.normalized) }))
        }
        var transaction = self
        transaction.memo = displayDescription ?? request?.displayDescription
        return transaction
    }
}
